src/commands/keygen.rs

```diff
@@ -0,0 +1,214 @@
+    #[arg(value_name = "domain name", value_parser = ValueParser::new(parse_name))]
+    name: Name<Vec<u8>>,
+}
+
+impl Keygen {
+    pub fn execute(self) -> Result<(), Error> {
+        // Determine the appropriate key generation parameters.
+        let params = match self.algorithm {
+            AlgorithmArg::List => {
+                // Print the algorithm list and exit.
+                println!("Possible algorithms:");
+                println!("  - RSASHA256 (8)");
+                println!("  - ECDSAP256SHA256 (13)");
+                println!("  - ECDSAP384SHA384 (14)");
+                println!("  - ED25519 (15)");
+                println!("  - ED448 (16)");
+                return Ok(());
+            }
+
+            AlgorithmArg::RsaSha256 => GenerateParams::RsaSha256 { bits: self.bits },
+            AlgorithmArg::EcdsaP256Sha256 => GenerateParams::EcdsaP256Sha256,
+            AlgorithmArg::EcdsaP384Sha384 => GenerateParams::EcdsaP384Sha384,
+            AlgorithmArg::Ed25519 => GenerateParams::Ed25519,
+            AlgorithmArg::Ed448 => GenerateParams::Ed448,
+        };
+
+        // The digest algorithm is selected based on the key algorithm.
+        let digest_alg = match params.algorithm() {
+            SecAlg::RSASHA256 => DigestAlg::SHA256,
+            SecAlg::ECDSAP256SHA256 => DigestAlg::SHA256,
+            SecAlg::ECDSAP384SHA384 => DigestAlg::SHA384,
+            SecAlg::ED25519 => DigestAlg::SHA256,
+            SecAlg::ED448 => DigestAlg::SHA256,
+            _ => unreachable!(),
+        };
+
+        // Generate the key.
+        // TODO: Attempt repeated generation to avoid key tag collisions.
+        let (secret_key, public_key) = common::generate(params)
+            .map_err(|err| format!("an implementation error occurred: {err}"))?;
+        let flags = if self.make_ksk { 257 } else { 256 };
+        let public_key = Key::new(self.name.clone(), flags, public_key);
+        let digest = self
+            .make_ksk
+            .then(|| public_key.digest(digest_alg).unwrap());
+
+        // Open the appropriate files to write the key.
+        let base = format!(
+            "K{}+{:03}+{:05}",
+            self.name.fmt_with_dot(),
+            public_key.algorithm().to_int(),
+            public_key.key_tag()
+        );
+        let mut secret_key_file = File::create_new(format!("{base}.private"))
+            .map_err(|err| format!("private key file '{base}.private' already existed: {err}"))?;
+        let mut public_key_file = File::create_new(format!("{base}.key"))
+            .map_err(|err| format!("public key file '{base}.key' already existed: {err}"))?;
+        let mut digest_file = self
+            .make_ksk
+            .then(|| File::create_new(format!("{base}.ds")))
+            .transpose()
+            .map_err(|err| format!("digest file '{base}.ds' already existed: {err}"))?;
+        #[cfg(target_family = "unix")]
+        if self.create_symlinks {
```

Comment:
ldns-keygen also creates a symlink for the .ds file if the key is a ksk

Reply:
I think I implemented that just a little before you reviewed here :)